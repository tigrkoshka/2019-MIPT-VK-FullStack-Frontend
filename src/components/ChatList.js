import React from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import Header from './Header'
import burger from '../images/burger.png'
import search from '../images/search.png'
import profilePic from '../images/profilePic.jpeg'
import newChat from '../images/new-chat.png'
import chatStyles from '../styles/singleChatStyles.module.scss'
import chatListStyles from '../styles/chatListStyles.module.scss'

const data = [
  'Общество целых бокалов',
  'Дженнифер Эшли',
  'Антон Иванов',
  'Серёга (должен 2000)',
  'Общество разбитых бокалов',
  'Сэм с Нижнего',
  'Айрат работа',
  'Кеша армия',
  'Первый курс ФПМИ-Наука 2019-2020',
]

function SingleChat({ name, lastTime, lastMessage, indicator, key }) {
  return (
    <div key={key}>
      <Link to={`MessageForm/${name}`}>
        <div className={chatStyles.chat}>
          <table>
            <tbody>
              <tr>
                <td rowSpan="2" className={chatStyles.for_pic}>
                  <div className={chatStyles.photo}>
                    <img src={profilePic} alt="" className={chatStyles.pic} height="100px" width="height" />
                  </div>
                </td>
                <td className={chatStyles.first}>
                  <div className={chatStyles.name}>{name}</div>
                  <div className={chatStyles.last_time}>{lastTime}</div>
                </td>
              </tr>
              <tr>
                <td className={chatStyles.second}>
                  <div className={chatStyles.last_message}>{lastMessage}</div>
                  <div className={chatStyles.indicator}>{indicator}</div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </Link>
      <hr />
    </div>
  )
}

SingleChat.propTypes = {
  name: PropTypes.string.isRequired,
  lastTime: PropTypes.string.isRequired,
  lastMessage: PropTypes.string.isRequired,
  indicator: PropTypes.string.isRequired,
  key: PropTypes.number.isRequired,
}

function setAll(name) {
  const props = {}

  props.name = name

  const messagesOfThisChat = JSON.parse(localStorage.getItem(name)) || []

  if (messagesOfThisChat.length !== 0) {
    const numberOfMessages = messagesOfThisChat.length

    props.lastTime = messagesOfThisChat[numberOfMessages - 1].time

    const preLastMess = messagesOfThisChat[0].content
    const arr = preLastMess.split(' ')
    let i = 0
    let count = arr[i].length
    let res = ''
    while (i < arr.length && count < 100) {
      res += `${arr[i]} `
      i += 1
      if (i < arr.length) {
        count += arr[i].length
      }
    }
    res = res.substr(0, res.length - 1)
    if (i < arr.length) {
      res += '...'
    }

    props.lastMessage = res
    props.indicator = numberOfMessages
  } else {
    props.lastTime = ''
    props.lastMessage = ''
    props.indicator = 0
  }

  return props
}

class ChatList extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      chats: [],
    }
    data.sort(this.compareNumber)
    for (let i = 0; i < data.length; i += 1) {
      const currProps = setAll(data[i])
      currProps.key = i
      const currChat = SingleChat(currProps)
      this.state.chats.push(currChat)
    }
  }

  compareNumber(a, b) {
    if (localStorage.getItem(a) === null) {
      if (localStorage.getItem(b) === null) {
        return 0
      }
      return 1
    }
    if (localStorage.getItem(b) === null) {
      return -1
    }
    const aLen = JSON.parse(localStorage.getItem(a)).length
    const bLen = JSON.parse(localStorage.getItem(b)).length
    if (aLen > bLen) {
      return -1
    }
    if (aLen === bLen) {
      return 0
    }
    if (aLen < bLen) {
      return 1
    }
    return 0
  }

  render() {
    return (
      <div className={chatListStyles.container}>
        <Header leftImg={burger} rightImg={search} leftLink="/UserProfile" name="Hummingbird" onRightClick={() => {}} />
        <div className={chatListStyles.chats}>{this.state.chats}</div>
        <div className={chatListStyles.new_chat}>
          <img src={newChat} height="70px" alt="" />
        </div>
      </div>
    )
  }
}

export default ChatList
