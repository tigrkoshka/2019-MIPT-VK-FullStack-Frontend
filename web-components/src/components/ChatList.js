const template = document.createElement('template');
template.innerHTML = `
    <style>
        .container {
            height: 100%;
            display: flex;
            flex-direction: column;
            transition: transform 2s;
        }
        
        #header {
            display: flex;
            flex-basis: 8%;
        }
        
        .chats {
            overflow-y: scroll;
            display: flex;
            flex-direction: column;
            width: 100%;
            height: 100%;
            cursor: pointer;
        }
        .new-chat {
            position: absolute;
            bottom: 30px;
            right: 30px;
        }
    </style>
    <div class="container"> 
        <chat-list-header id="header"></chat-list-header>
        <div class="chats"></div>
        <div class="new-chat"><img src="/src/components/images/new-chat.png" height="70px" alt=""></div>
    </div>
`;

class ChatList extends HTMLElement {
  constructor() {
    super();
    this._shadowRoot = this.attachShadow({ mode: 'open' });
    this._shadowRoot.appendChild(template.content.cloneNode(true));

    this.$chatsList = this._shadowRoot.querySelector('.chats');
    this.$header = this._shadowRoot.querySelector('chat-list-header');

    const data = ['Общество целых бокалов',
      'Дженнифер Эшли',
      'Антон Иванов',
      'Серёга (должен 2000)',
      'Общество разбитых бокалов',
      'Сэм с Нижнего',
      'Айрат работа',
      'Кеша армия',
      'Первый курс ФПМИ-Наука 2019-2020'];

    data.sort(this.compareNumber);

    for (let i = data.length - 1; i >= 0; i -= 1) {
      if (localStorage.getItem(data[i]) === null) {
        const a = [];
        localStorage.setItem(data[i], JSON.stringify(a));
      }

      const currChat = document.createElement('one-chat');
      currChat.setAll(data[i]);

      this.$chatsList.insertBefore(currChat, this.$chatsList.firstChild);
    }

    this.$chats = this._shadowRoot.querySelectorAll('one-chat');
  }

  get header() {
    return this.$header;
  }

  get chats() {
    return this.$chats;
  }

  get container() {
    return this._shadowRoot.querySelector('.container');
  }

  compareNumber(a, b) {
    if (localStorage.getItem(a) === null) {
      if (localStorage.getItem(b) === null) {
        return 0;
      }
      return 1;
    }
    if (localStorage.getItem(b) === null) {
      return -1;
    }
    const aLen = JSON.parse(localStorage.getItem(a)).length;
    const bLen = JSON.parse(localStorage.getItem(b)).length;
    if (aLen > bLen) {
      return -1;
    }
    if (aLen === bLen) {
      return 0;
    }
    if (aLen < bLen) {
      return 1;
    }
    return 0;
  }
}

customElements.define('chat-list', ChatList);
