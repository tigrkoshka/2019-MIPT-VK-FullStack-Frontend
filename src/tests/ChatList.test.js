/* eslint-disable react/jsx-props-no-spreading */
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import ChatList from '../components/ChatList'

describe('Chat list renders and acts correctly', () => {
  let props
  let component

  beforeEach(() => {
    props = {
      match: {
        params: {
          userId: '1',
        },
      },
    }

    component = renderer
      .create(
        <MemoryRouter>
          <ChatList {...props} />
        </MemoryRouter>,
      )
      .toJSON()
  })

  it('Should render normally', () => {
    expect(component).toMatchSnapshot()
  })
})
