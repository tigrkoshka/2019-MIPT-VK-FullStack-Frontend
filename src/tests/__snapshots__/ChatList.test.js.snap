// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat list renders and acts correctly Should render normally 1`] = `
<div
  className="container"
>
  <div
    className="rectangle"
  >
    <div
      className="leftHorizontal"
    >
      <a
        className="leftButton"
        href="/UserProfile/1"
        onClick={[Function]}
      >
        <div
          className="vertical"
        >
          <img
            alt=""
            height="25px"
            src="burger.png"
            width="height"
          />
        </div>
      </a>
    </div>
    <div
      className="horizontal"
    >
      <div
        className="vertical"
      >
        <div
          className="name"
        >
          Hummingbird
        </div>
      </div>
    </div>
    <div
      className="rightHorizontal"
    >
      <div
        className="rightButton"
        onClick={[Function]}
      >
        <div
          className="vertical"
        >
          <div
            className="text"
          >
            Exit
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="chats"
  />
  <a
    className="new_chat"
    href="/CreateChat/1"
    onClick={[Function]}
  >
    <img
      alt=""
      height="70px"
      src="new-chat.png"
    />
  </a>
</div>
`;
